rpm.8
.\" rpm - Red Hat Package Manager
.TH rpm 8 "15 July 1996" "Red Hat Software" "Red Hat Linux"
.SH NAME
rpm \- Red Hat Package Manager
.SH SYNOPSIS
\fBrpm\fP [options] 
.SH DESCRIPTION
\fBrpm\fP is a powerful \fIpackage manager\fP, which can be used to
build, install, query, verify, update, and uninstall individual
software packages.  A \fIpackage\fP consists of an archive of files,
and package information, including name, version, and description.

There are ten basic modes of operation, and each takes a different
set of options.  They are \fIInstall\fP, \fIQuery\fP,
\fIVerify\fP, \fISignature check\fP, \fIUninstall\fP, \fIBuild\fP,
\fIRebuild Database\fP, \fIfix permissions\fP, \fIset owners and groups\fR, 
and \fIShow RC\fP.

Install mode:
.br
.I "\fB    rpm \-i [install\-options] <package_file>+\fP"
.br
Query mode:
.br
.I "\fB    rpm \-q [query\-options]\fP"
.br
Verify mode:
.br
.I "\fB    rpm \-V|\-y|\-\-verify [verify\-options]\fP"
.br
Signature Check mode:
.br
.I "\fB    rpm \-\-checksig <package_file>+\fP"
.br
Uninstall mode:
.br
.I "\fB    rpm \-e <package_name>+\fP"
.br
Build mode:
.br
.I "\fB    rpm \-[b|t]\fIO\fB [build\-options] <package_spec>+\fP"
.br
Rebuild database:
.br
.I "\fB    rpm \-\-rebuilddb"
.br
Fix permissions:
.br
.I "\fB    rpm \-\-setperms [query\-package\-specifiers]\fP"
.br
Set owners and groups:
.br
.I "\fB    rpm \-\-setugids [query\-package\-specifiers]\fP"
.br
Show RC:
.br
.I "\fB    rpm \-\-showrc"
.br

.SH GENERAL OPTIONS
These options can be used in all the different modes.
.IP "\fB\-vv\fP"
Print lots of ugly debugging information.
.IP "\fB\-\-keep\-temps\fP"
Do not remove temporary files (/tmp/rpm\-*).  Primarily only useful
for debugging rpm.
.IP "\fB\-\-quiet\fP"
Print as little as possible \- normally only error messages will be
displayed.
.IP "\fB\-\-help\fP"
Print a longer usage message then normal.
.IP "\fB\-\-version\fP"
Print a single line containing the version number of rpm being used.
.IP "\fB\-\-rcfile <file>\fP"
Use \fB<file>\fP instead of \fB/etc/rpmrc\fP and \fB$HOME/.rpmrc\fP.
.IP "\fB\-\-root <dir>\fP"
Use the directory given as top level directory for all operations.

.SH INSTALL OPTIONS
The general form of an rpm install command is
.PP
	\fBrpm \-i [install\-options] <package_file>+\fP
.PP
The <package_file> may be specified as an ftp style URL, in which case
the package will be downloaded before being installed. See \fBFTP
OPTIONS\fP for information on RPM's built in ftp support.
.PP
.IP "\fB\-\-force\fP"
Same as using both \-\-replacepkgs, \-\-replacefiles, and 
\-\-oldpackage.
.IP "\fB\-h, \-\-hash\fP"
Print 50 hash marks as the package archive is unpacked.  Use
with \fB\-v\fP for a nice display.
.IP "\fB\-\-oldpackage\fP"
Allow an upgrade to replace a newer package with an older one.
.IP "\fB\-\-percent\fP"
Print percentages as files are unpacked from the package archive. This
is intended to make RPM easy to run from other tools.
.IP "\fB\-\-replacefiles\fP"
Install the packages even if they replace files from other, already
installed, packages.
.IP "\fB\-\-replacepkgs\fP"
Install the packages even if some of them are already installed on this
system.
.IP "\fB\-\-root \fI<dir>\fP"
Do the installation on the system rooted at \fI<dir>\fP.  Note that
this means the database will be updated under \fI<dir>\fP and any
\fIpre\fP or \fIpost\fP scripts are run after a chroot() to \fI<dir>\fP.
.IP "\fB\-\-nodeps\fP"
Don't do a dependency check before installing the new package.
.IP "\fB\-\-noscripts\fP"
Don't execute the preinstall or postinstall scripts.
.IP "\fB\-\-excludedocs\fP"
Don't install any files which are marked as documentation (which includes
man pages and texinfo documents).
.IP "\fB\-\-includedocs\fP"
Install documentation files. This is only needed if \fIexcludedocs: 1\fP
is specified in an rpmrc file.
.IP "\fB\-\-nodeps\fP"
Don't check dependencies before installing the packages.
.IP "\fB\-\-test\fP"
Do not install the package, simply check for and report potential
conflicts.
.IP "\fB\-U, \-\-upgrade\fP"
Upgrade the package currently installed to the version in the new RPM.
This is the same as install, except all other version of the package
are removed from the system.



.SH QUERY OPTIONS
The general form of an rpm query command is
.PP
	\fBrpm \-q [query\-options]\fP
.PP
You may specify the format that package information should be printed 
in. To do this, you use the \fB\-\-queryformat\fP option, followed by 
the format string.

Query formats are modifed versions of the standard \fBprintf(3)\fP
formatting. The format is made up of static strings (which may include
standard C character escapes for newlines, tabs, and other special
characters) and \fBprintf(3)\fP type formatters. As \fBrpm\fP
already knows the type to print, the type specifier must be
omitted however, and replaced by the name of the header tag to
be printed, enclosed by \fB{}\fP characters. The 
\fBRPMTAG_\fP portion of the tag name may be omitted, and the
tag name may be preceded by a \fB\-\fP to format the tag as
a date if possible.

For example, to print only the names of the packages queried, you
could use \fB%{NAME}\fP as the format string. To print the packages
name and distribution information in two columns, you could use
\fB%\-30{NAME}%{DISTRIBUTION}\fP.

\fBrpm\fP will print a list of all of the tags it knows about when
it is invoked with the \fB\-\-querytags\fP argument.

There are two subsets of options for querying: package selection, and
information selection.

Package selection options:
.br
.IP "\fB\fI<package_name>\fP"
Query package named \fB\fI<package_name>\fP.
.IP "\fB\-a\fP"
Query all packages
.IP "\fB\-\-whatrequires \fI<capability>\fP"
Query all packages that requires \fI<capability>\fP for proper functioning.
.IP "\fB\-\-whatprovides \fI<virtual>\fP"
Query all packages that provide the \fI<virtual>\fP capability.
.IP "\fB\-f \fI<file>\fP"
Query package owning \fI<file>\fP.
.IP "\fB\-p \fI<package_file>\fP"
Query an (uninstalled) package \fI<package_file>\fP. 
The <package_file> may be specified as an ftp style URL, in which case
the package header will be downloaded and queried. See \fBFTP
OPTIONS\fP for information on RPM's built in ftp support.

.P
Information selection options:
.br
.IP "\fB\-i\fP"
Display package information, including name, version, and description. This
uses the \fB\-\-queryformat\fP if one was specified.
.IP "\fB\-R\fP"
List packages this one depends on (same as \fB\-\-requires\fP).
.IP "\fB\-\-provides\fP"
List capabilities this package provides.
.IP "\fB\-l\fP"
List files in package.
.IP "\fB\-s\fP"
Display the \fIstates\fP of files in the package (implies \fB\-l\fP).  
The state of each file is
either \fInormal\fP, \fInot installed\fP, or \fIreplaced\fP.
.IP "\fB\-d\fP"
List only documentation files (implies \fB\-l\fP).  
.IP "\fB\-c\fP"
List only configuration files (implies \fB\-l\fP).  
.IP "\fB\-\-scripts\fP"
List the package specific shell scripts that are used as part of the 
installation and uninstallation processes, if there are any.
.IP "\fB\-\-dump\fP"
Dump file information as follows: path size mtime md5sum mode
owner group isconfig isdoc rdev symlink. This must be used with
at least one of \fB\-l\fP, \fB\-c\fP, \fB\-d\fP.

.SH VERIFY OPTIONS
The general form of an rpm verify command is
.PP
	\fBrpm \-V|\-y|\-\-verify [verify\-options]\fP
.PP
Verifying a package compares information about the installed
files in the package with information about the files taken from the
original package and stored in the rpm database.  Among other things,
verifying compares the size, MD5 sum, permissions, type, owner and group
of each file.  Any discrepencies are displayed.  The package specification
options are the same as for package querying.

Files that were not installed from the package, for example documentation
files excluded on installation using the "\fB\-\-excludedocs\fP" option,
will be silently ignored.

The format of the output is a string of 8 characters, a possible "\fBc\fP"
denoting a configuration file, and then the file name.  Each of the 8
characters denotes the result of a comparison of one attribute of the file
to the value of that attribute recorded in the RPM database.  A 
single "\fB.\fP" (period) means the test passed.  The following characters 
denote failure of certain tests:

.IP "\fB5\fP"
MD5 sum
.IP "\fBS\fP"
File size
.IP "\fBL\fP"
Symlink
.IP "\fBT\fP"
Mtime
.IP "\fBD\fP"
Device
.IP "\fBU\fP"
User
.IP "\fBG\fP"
Group
.IP "\fBM\fP"
Mode (includes permissions and file type)

.SH SIGNATURE CHECKING
The general form of an rpm signature check command is
.PP
	\fBrpm \-\-checksig <package_file>+\fP
.PP
This checks the PGP signature built into a package to ensure the integrity
and the origin of the package.
PGP configuration information is read from /etc/rpmrc.
See the section on PGP SIGNATURES for details.

.SH UNINSTALL OPTIONS
The general form of an rpm uninstall command is
.PP
	\fB    rpm \-e <package_name>+\fP
.PP
.IP "\fB\-\-noscripts\fP"
Don't execute the preuninstall or postuninstall scripts.
.IP "\fB\-\-nodeps\fP"
Don't check dependencies before uninstalling the packages.
.IP "\fB\-\-test\fP"
Don't really uninstall anything, just go through the motions. 
\fB\-vv\fP option.
.IP "\fB\-\-nodeps\fP"
Don't check for broken dependencies before removing the package.

.SH BUILD OPTIONS
The general form of an rpm build command is
.PP
    \fBrpm \-[b|t]\fIO\fP [build\-options] <package_spec>+\fP
.PP
The argument used is \fB-b\fR if a spec file is being used to build the package
and \fB-t\fR if \fBRPM\fR should look inside of a gzipped (or compressed) tar
file for the spec file to use. After the first argument, the next argument
(\fIO\fR) specifies the stages of building and packaging to be done and
is one of:

.IP "\fB\-bp\fP"
Executes the "%prep" stage from the spec file.  Normally this
involves unpacking the sources and applying any patches.
.IP "\fB\-bl\fP"
Do a "list check".  The "%files" section from the spec file
is macro expanded, and checks are made to insure the files
exist.
.IP "\fB\-bc\fP"
Do the "%build" stage from the spec file (after doing the prep stage).
This generally involves the equivalent of a "make".
.IP "\fB\-bi\fP"
Do the "%install" stage from the spec file (after doing the prep
and build stages).  This generally involves the equivalent of a
"make install".
.IP "\fB\-bb\fP"
Build a binary package (after doing the prep, build, and install stages).
.IP "\fB\-ba\fP"
Build binary and source packages (after doing the prep, build, and 
install stages).
.PP
The following options may also be used:
.IP "\fB\-\-short\-circuit\fP"
Skip straight to specified stage (ie, skip all stages leading up
to the specified stage).  Only valid with \fB\-bc\fP and \fB\-bi\fP.
.IP "\fB\-\-timecheck\fP"
Set the "timecheck" age (0 to disable).  This value can also
be set in rpmrc with "timecheck:".  The timecheck value expresses,
in seconds, the maximum age of a file being packaged.  Warnings
will be printed for all files beyond the timecheck age.
.IP "\fB\-\-clean\fP"
Remove the build tree after the packages are made.
.IP "\fB\-\-test\fP"
Do not execute any build stages.
Useful for testing out spec files.
.IP "\fB\-\-sign\fP"
Embed a PGP signature in the package.  This signature can be used
to verify the integrity and the origin of the package.  See the
section on PGP SIGNATURES for /etc/rpmrc details.

.SH REBUILD AND RECOMPILE OPTIONS

There are two other ways to invoke rpm:

.I "\fBrpm \-\-recompile <source_package_file>+\fP"

.I "\fBrpm \-\-rebuild <source_package_file>+\fP"

When invoked this way, rpm installs the named source package, and does
a prep, compile and install.  In addition, \fB\-\-rebuild\fP builds a new
binary package. When the build has completed, the build directory is
removed (as in \fB\-\-clean\fP) and the the sources and spec file for
the package are removed.

.SH SIGNING AN EXISTING RPM

.I "\fBrpm \-\-resign <binary_package_file>+\fP"

This option generates and inserts new signatures for the listed packages.
Any existing signatures are removed.

.SH PGP SIGNATURES

In order to use the signature feature RPM must be able to run PGP
(it must be installed and in your path), and it must be able to
find a public key ring with RPM public keys in it.  By default,
RPM uses the PGP defaults to find the keyrings (honoring PGPPATH).
If your key rings are not located where PGP expects them to be,
you must set the following in your /etc/rpmrc

.IP "\fBpgp_path\fP"
Replacement path for /usr/lib/rpm.  Must contain your key rings.
.PP

If you want to be able to sign packages you create yourself, you also
need to create your own public and secret key pair (see the PGP manual).
In addition to the above /etc/rpmrc entries, you should add the following:

.IP "\fBsignature\fP"
The signature type.  Right now only pgp is supported.
.IP "\fBpgp_name\fP"
The name of the "user" whose key you wish to use to sign your packages.
.PP

When building packages you then add \-\-sign to the command line.
You will be prompted for your pass phrase, and your package will
be built and signed.

.SH REBUILD DATABASE OPTIONS

The general form of an rpm rebuild database command is
.PP
	\fBrpm \-\-rebuilddb\fP
.PP

The only options this mode supports are \fB-\-dbpath\fP and \fB-\-root\fP.

.SH SHOWRC 

Running 

.PP
	\fBrpm \-\-showrc\fP

.PP
shows the values RPM will use for all of the options that may be set
in \fIrpmrc\fP files.

.SH FTP OPTIONS

RPM includes a simple FTP client to simplify installing and querying packages 
which are available over the internet. Package files for install, upgrade,
and query operations may be specified as an ftp style URL:

.PP
	\fBftp://<user>@<password>:hostname/path/to/package.rpm\fP
.PP
If the \fB@password\fP portion is omitted, the password will be prompted
for (once per user/hostname pair). If both the user and password are
omitted, anonymous ftp is used. In all cases passive (PASV) ftp transfers are
used.

RPM allows the folowing options to be used with ftp URLs:

.IP "\fB\--ftpproxy <hostname>\fP"
The host <hostname> will be used as a proxy server for all transfers, which
allows users to ftp through firewall machines which use proxy systems.  This
option may also be specified in an \fIrpmrc\fP file.

.IP "\fB\--ftpport <port>\fP"
Specifies the TCP port number to use for the ftp connection instead of
the default port.
This option may also be specified in an \fIrpmrc\fP file.

.SH FILES
.nf
/etc/rpmrc
~/.rpmrc
/var/lib/rpm/packages
/var/lib/rpm/pathidx
/var/lib/rpm/nameidx
/tmp/rpm*
.fi
.El
.SH SEE ALSO
.IR glint (8) ,
.IR rpm2cpio (8) ,
.B http://www.redhat.com/rpm
.nf
.SH AUTHORS
.nf
Marc Ewing <marc@redhat.com>
Erik Troan <ewt@redhat.com>
.fi


NEWS
Not yet implemented in 2.0:

. dependencies
. list check (-bl)
. --where and --search

New in RPM 2.0 (or previously undocumented):

*** Install

--replacepkgs	- will install a package even if it's already in the database
--replacefiles  - will install a package even if the files it contains conflict
		  with others
--force		- same as "--replacepkgs --replacefiles"

*** Uninstall

*** Upgrade

*** Query

*** Verify

Verification output is much more readable.  Notes config files, etc.

*** PGP Signatures

RPM now supports optional PGP signatures embedded in RPM packages.
Only 1024 bit keys are supported, and the signature "covers" the
header and archive sections of the package.  The "lead", which 
makes up the first 96 bytes of every package is not covered by
the signature.  This should not be cause for alarm as the lead
is only used by file(1), and by RPM to determine the package type
(either source or binary) and architecture.

The rpm(1) man page has some details on using PGP signatures.

*** Spec file (RPM building)

. Summary: ...
  A one (short) line description of the package.  This field
  takes over what Description used to do, but is limited to
  a single line.  For an interim period, if the Summary is
  not defined the description will be used instead.

. Description: ...
  An optional mulit-line description of the package.

. Packager: ...
  Name and email address of the person who "maintains" the RPM

. URL: ...
  A place to put a URL for more info on the package (http://...)

. Serial: ...
  This number should be monotonically increasing with each new
  release of the package.  If present, it will be used by RPM
  to define an ordering.  If not present RPM tries to determine
  the ordering from the Version field.

. %package -n foo
  Builds "subpackage" with name foo-Version-Release

. %ifarch <arch>+
  %ifnarch <arch>+
  %else
  %endif
  also: %ifos <os>+
        %ifnos <os>+

. Exclude: <arch>+
  Exclusive: <arch>+

. %clean
  Executed after all packaging is finished

. %changelog

. %verify can be used to control which aspects of the file are verified

. %attr can be used to specify mode, user and group of files

. %files -f <file>
  Takes file list from <file>, which is relative to the build dir.


tools/rpm.c
#include "system.h"

#include <rpm/rpmcli.h>
#include <rpm/rpmlib.h>			/* RPMSIGTAG, rpmReadPackageFile .. */
#include <rpm/rpmlog.h>
#include <rpm/rpmps.h>
#include <rpm/rpmstring.h>
#include <rpm/rpmts.h>

#include "cliutils.h"

#include "debug.h"

enum modes {

    MODE_QUERY		= (1 <<  0),
    MODE_VERIFY		= (1 <<  3),
#define	MODES_QV (MODE_QUERY | MODE_VERIFY)

    MODE_INSTALL	= (1 <<  1),
    MODE_ERASE		= (1 <<  2),
    MODE_RESTORE	= (1 <<  4),
#define	MODES_IE (MODE_INSTALL | MODE_ERASE)

    MODE_UNKNOWN	= 0
};

#define	MODES_FOR_NODEPS	(MODES_IE | MODE_VERIFY)
#define	MODES_FOR_TEST		(MODES_IE)

static int quiet;

/* the structure describing the options we take and the defaults */
static struct poptOption optionsTable[] = {

 { NULL, '\0', POPT_ARG_INCLUDE_TABLE, rpmQVSourcePoptTable, 0,
        N_("Query/Verify package selection options:"),
        NULL },
 { NULL, '\0', POPT_ARG_INCLUDE_TABLE, rpmQVFilePoptTable, 0,
        N_("Query/Verify file selection options:"),
        NULL },
 { NULL, '\0', POPT_ARG_INCLUDE_TABLE, rpmQueryPoptTable, 0,
	N_("Query options (with -q or --query):"),
	NULL },
 { NULL, '\0', POPT_ARG_INCLUDE_TABLE, rpmVerifyPoptTable, 0,
	N_("Verify options (with -V or --verify):"),
	NULL },
 { NULL, '\0', POPT_ARG_INCLUDE_TABLE, rpmInstallPoptTable, 0,
	N_("Install/Upgrade/Erase options:"),
	NULL },

 { "quiet", '\0', POPT_ARGFLAG_DOC_HIDDEN, &quiet, 0, NULL, NULL},

 { NULL, '\0', POPT_ARG_INCLUDE_TABLE, rpmcliAllPoptTable, 0,
	N_("Common options for all rpm modes and executables:"),
	NULL },

   POPT_AUTOALIAS
   POPT_AUTOHELP
   POPT_TABLEEND
};

int main(int argc, char *argv[])
{
    rpmts ts = NULL;
    enum modes bigMode = MODE_UNKNOWN;
    QVA_t qva = &rpmQVKArgs;
    struct rpmInstallArguments_s * ia = &rpmIArgs;
    poptContext optCon;
    int ec = 0;
    int i;

    optCon = rpmcliInit(argc, argv, optionsTable);

    /* Set the major mode based on argv[0] */
    if (rstreq(xgetprogname(), "rpmquery"))	bigMode = MODE_QUERY;
    if (rstreq(xgetprogname(), "rpmverify")) bigMode = MODE_VERIFY;

    /* Jumpstart option from argv[0] if necessary. */
    switch (bigMode) {
    case MODE_QUERY:	qva->qva_mode = 'q';	break;
    case MODE_VERIFY:	qva->qva_mode = 'V';	break;
    case MODE_INSTALL:
    case MODE_ERASE:
    case MODE_UNKNOWN:
    default:
	break;
    }

  if (bigMode == MODE_UNKNOWN || (bigMode & MODES_QV)) {
    switch (qva->qva_mode) {
    case 'q':	bigMode = MODE_QUERY;		break;
    case 'V':	bigMode = MODE_VERIFY;		break;
    }

    if (qva->qva_sourceCount) {
	if (qva->qva_sourceCount > 1)
	    argerror(_("one type of query/verify may be performed at a "
			"time"));
    }
    if (qva->qva_flags && (bigMode & ~MODES_QV)) 
	argerror(_("unexpected query flags"));

    if (qva->qva_queryFormat && (bigMode & ~MODES_QV)) 
	argerror(_("unexpected query format"));

    if (qva->qva_source != RPMQV_PACKAGE && (bigMode & ~MODES_QV)) 
	argerror(_("unexpected query source"));
  }

  if (bigMode == MODE_UNKNOWN || (bigMode & MODES_IE))
    {	int iflags = (ia->installInterfaceFlags &
			(INSTALL_UPGRADE|INSTALL_FRESHEN|
			 INSTALL_INSTALL|INSTALL_REINSTALL));
	int eflags = (ia->installInterfaceFlags & INSTALL_ERASE);
	int rflags = (ia->installInterfaceFlags & INSTALL_RESTORE);

	if (iflags & eflags & rflags)
	    argerror(_("only one major mode may be specified"));
	else if (iflags)
	    bigMode = MODE_INSTALL;
	else if (eflags)
	    bigMode = MODE_ERASE;
	else if (rflags)
	    bigMode = MODE_RESTORE;
    }

    if (!( bigMode == MODE_INSTALL ) &&
(ia->probFilter & (RPMPROB_FILTER_REPLACEPKG | RPMPROB_FILTER_OLDPACKAGE)))
	argerror(_("only installation and upgrading may be forced"));
    if (bigMode != MODE_INSTALL && (ia->probFilter & RPMPROB_FILTER_FORCERELOCATE))
	argerror(_("files may only be relocated during package installation"));

    if (ia->relocations && ia->prefix)
	argerror(_("cannot use --prefix with --relocate or --excludepath"));

    if (bigMode != MODE_INSTALL && ia->relocations)
	argerror(_("--relocate and --excludepath may only be used when installing new packages"));

    if (bigMode != MODE_INSTALL && ia->prefix)
	argerror(_("--prefix may only be used when installing new packages"));

    if (ia->prefix && ia->prefix[0] != '/') 
	argerror(_("arguments to --prefix must begin with a /"));

    if (!(bigMode & MODES_IE) && (ia->installInterfaceFlags & INSTALL_HASH))
	argerror(_("--hash (-h) may only be specified during package "
			"installation and erasure"));

    if (!(bigMode & MODES_IE) && (ia->installInterfaceFlags & INSTALL_PERCENT))
	argerror(_("--percent may only be specified during package "
			"installation and erasure"));

    if (bigMode != MODE_INSTALL && (ia->probFilter & RPMPROB_FILTER_REPLACEPKG))
	argerror(_("--replacepkgs may only be specified during package "
			"installation"));

    if (bigMode != MODE_INSTALL && (ia->transFlags & RPMTRANS_FLAG_NODOCS))
	argerror(_("--excludedocs may only be specified during package "
		   "installation"));

    if (bigMode != MODE_INSTALL && ia->incldocs)
	argerror(_("--includedocs may only be specified during package "
		   "installation"));

    if (ia->incldocs && (ia->transFlags & RPMTRANS_FLAG_NODOCS))
	argerror(_("only one of --excludedocs and --includedocs may be "
		 "specified"));
  
    if (bigMode != MODE_INSTALL && (ia->probFilter & RPMPROB_FILTER_IGNOREARCH))
	argerror(_("--ignorearch may only be specified during package "
		   "installation"));

    if (bigMode != MODE_INSTALL && (ia->probFilter & RPMPROB_FILTER_IGNOREOS))
	argerror(_("--ignoreos may only be specified during package "
		   "installation"));

    if (bigMode != MODE_INSTALL && bigMode != MODE_ERASE &&
	(ia->probFilter & (RPMPROB_FILTER_DISKSPACE|RPMPROB_FILTER_DISKNODES)))
	argerror(_("--ignoresize may only be specified during package "
		   "installation"));

    if ((ia->installInterfaceFlags & UNINSTALL_ALLMATCHES) && bigMode != MODE_ERASE)
	argerror(_("--allmatches may only be specified during package "
		   "erasure"));

    if ((ia->transFlags & RPMTRANS_FLAG_ALLFILES) && bigMode != MODE_INSTALL)
	argerror(_("--allfiles may only be specified during package "
		   "installation"));

    if ((ia->transFlags & RPMTRANS_FLAG_JUSTDB) &&
	bigMode != MODE_INSTALL && bigMode != MODE_ERASE)
	argerror(_("--justdb may only be specified during package "
		   "installation and erasure"));

    if (bigMode != MODE_INSTALL && bigMode != MODE_ERASE && bigMode != MODE_VERIFY &&
	(ia->transFlags & (RPMTRANS_FLAG_NOSCRIPTS | _noTransScripts | _noTransTriggers)))
	argerror(_("script disabling options may only be specified during "
		   "package installation and erasure"));

    if (bigMode != MODE_INSTALL && bigMode != MODE_ERASE && bigMode != MODE_VERIFY &&
	(ia->transFlags & (RPMTRANS_FLAG_NOTRIGGERS | _noTransTriggers)))
	argerror(_("trigger disabling options may only be specified during "
		   "package installation and erasure"));

    if (ia->noDeps & (bigMode & ~MODES_FOR_NODEPS))
	argerror(_("--nodeps may only be specified during package "
		   "installation, erasure, and verification"));

    if ((ia->transFlags & RPMTRANS_FLAG_TEST) && (bigMode & ~MODES_FOR_TEST))
	argerror(_("--test may only be specified during package installation "
		 "and erasure"));

    if (rpmcliRootDir && rpmcliRootDir[0] != '/') {
	argerror(_("arguments to --root (-r) must begin with a /"));
    }

    if (quiet)
	rpmSetVerbosity(RPMLOG_WARNING);

    if (rpmcliPipeOutput && initPipe())
	exit(EXIT_FAILURE);
	
    ts = rpmtsCreate();
    (void) rpmtsSetRootDir(ts, rpmcliRootDir);
    switch (bigMode) {
    case MODE_ERASE:
	if (ia->noDeps) ia->installInterfaceFlags |= UNINSTALL_NODEPS;

	if (!poptPeekArg(optCon)) {
	    argerror(_("no packages given for erase"));
	} else {
	    ec += rpmErase(ts, ia, (ARGV_const_t) poptGetArgs(optCon));
	}
	break;

    case MODE_INSTALL:

	/* RPMTRANS_FLAG_KEEPOBSOLETE */

	if (!ia->incldocs) {
	    if (ia->transFlags & RPMTRANS_FLAG_NODOCS) {
		;
	    } else if (rpmExpandNumeric("%{_excludedocs}"))
		ia->transFlags |= RPMTRANS_FLAG_NODOCS;
	}

	if (ia->noDeps) ia->installInterfaceFlags |= INSTALL_NODEPS;

	/* we've already ensured !(!ia->prefix && !ia->relocations) */
	if (ia->prefix) {
	    ia->relocations = (struct rpmRelocation_s *)xmalloc(2 * sizeof(*ia->relocations));
	    ia->relocations[0].oldPath = NULL;   /* special case magic */
	    ia->relocations[0].newPath = ia->prefix;
	    ia->relocations[1].oldPath = NULL;
	    ia->relocations[1].newPath = NULL;
	} else if (ia->relocations) {
	    ia->relocations = xrealloc(ia->relocations, 
			sizeof(*ia->relocations) * (ia->numRelocations + 1));
	    ia->relocations[ia->numRelocations].oldPath = NULL;
	    ia->relocations[ia->numRelocations].newPath = NULL;
	}

	if (!poptPeekArg(optCon)) {
	    argerror(_("no packages given for install"));
	} else {
	    /* FIX: ia->relocations[0].newPath undefined */
	    ec += rpmInstall(ts, ia, (ARGV_t) poptGetArgs(optCon));
	}
	break;

    case MODE_RESTORE:
	ec += rpmRestore(ts, ia, (ARGV_const_t) poptGetArgs(optCon));
	break;

    case MODE_QUERY:
	if (!poptPeekArg(optCon) && !(qva->qva_source == RPMQV_ALL))
	    argerror(_("no arguments given for query"));

	ec = rpmcliQuery(ts, qva, (ARGV_const_t) poptGetArgs(optCon));
	break;

    case MODE_VERIFY:
    {	rpmVerifyFlags verifyFlags = VERIFY_ALL;

	verifyFlags &= ~qva->qva_flags;
	qva->qva_flags = (rpmQueryFlags) verifyFlags;

	rpmtsSetFlags(ts, rpmtsFlags(ts) | (ia->transFlags & RPMTRANS_FLAG_NOPLUGINS));

	if (!poptPeekArg(optCon) && !(qva->qva_source == RPMQV_ALL))
	    argerror(_("no arguments given for verify"));
	ec = rpmcliVerify(ts, qva, (ARGV_const_t) poptGetArgs(optCon));
    }	break;

    case MODE_UNKNOWN:
	if (poptPeekArg(optCon) != NULL || argc <= 1) {
	    printUsage(optCon, stderr, 0);
	    ec = argc;
	}
	break;
    }

    rpmtsFree(ts);
    if (finishPipe())
	ec = EXIT_FAILURE;

    free(qva->qva_queryFormat);

    if (ia->relocations != NULL) {
	for (i = 0; i < ia->numRelocations; i++)
	    free(ia->relocations[i].oldPath);
	free(ia->relocations);
    }

    rpmcliFini(optCon);

    rpmlog(RPMLOG_DEBUG, "Exit status: %d\n", ec);

    return RETVAL(ec);
}
